File: Microverse Challenges/module 4/4.2 depth-first-search-Trast00/challenge.rb
def depth_first_search(graph)
  # write your code here
  @result = [0]
  @discovered = [0]
  @graph = graph

  def depth(key)
    @result << key if !@result.include?(key)

    list_nodes = @graph[key]
    list_nodes.each do |node|
      if !@discovered.include?(node)
        @discovered << node
        depth(node)
      end
    end
  end

  depth(0)

  @result
end

p depth_first_search({
  0 => [2], 
  1 => [4], 
  2 => [5, 0, 3], 
  3 => [2], 
  4 => [1, 5], 
  5 => [4, 2]
})
# => [0, 2, 5, 4, 1, 3]

File: Microverse Challenges/module 2/2.3 queue-Trast00/challenge.js
// Start with your code from LinkedList challenge.
class Node {
  constructor(value, next_node = null) {
    this.value = value;
    this.next_node = next_node;
  }
}

class Queue {
  head = null;
  add(number) {
    // your code here
    if (this.head === null){
      this.head = new Node(number);
      return number;
    } else {
      var current = this.head
      while (current.next_node != null){
        current = current.next_node
      }
      current.next_node = new Node(number)
    }
  }
  
  remove() {
    // your code here
    if(this.head === null) {
      return -1;
    }else {
      let result = this.head.value
      this.head = this.head.next_node
      return result
    }
  }
}

const queue = new Queue()

queue.add(3)
queue.add(5)
console.log(queue.remove())
// => 3

queue.add(2)
queue.add(7)
console.log(queue.remove())
// => 5

console.log(queue.remove())
// => 2

console.log(queue.remove())
// => 7

console.log(queue.remove())
// => -1

module.exports = Queue
